atlp-rwanda/ecommerce-app-predators-bn
# #4 feature apply coupons during checkout

#### What does this PR do?

This PR adds backend functionality for buyers to apply coupons during checkout. The backend has been updated to handle the application of coupons and ensure that the discount is accurately applied to the order total.

#### Description of Task to be completed?

The task was to implement coupon functionality in the backend, such that given a valid coupon code and a logged-in user, the discount should be applied to the order total and a confirmation message should be sent to the frontend. If the coupon code is invalid or the user is not logged in, an appropriate error message should be sent to the frontend. The backend should also handle scenarios where the coupon code has already been used or has expired.

#### How should this be manually tested?

1. Start the backend server by running `npm run start`.
2. Send a POST request to the `/checkout/apply-coupon` endpoint with a valid coupon code and a logged-in user.
3. Verify that the response includes the updated order total with the discount applied and a confirmation message is included.
4. Send a POST request to the `/checkout/apply-coupon` endpoint with an invalid coupon code or without a logged-in user.
5. Verify that the response includes an appropriate error message.
6. Send a POST request to the `/checkout/apply-coupon` endpoint with a coupon code that has already been used or has expired.
7. Verify that the response includes an appropriate error message.

#### Any background context you want to provide?

This feature was requested by the client to provide buyers with the ability to apply coupons during checkout.

#### What are the relevant pivotal tracker/Trello stories?

This PR is related to the following Trello story: https://trello.com/c/9ea28RJF

#### Screenshots (if appropriate)

N/A

#### Questions:

N/A
